presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.planner.DependencyExtractor;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Optional;
+
+public class MergeAdjacentWindows
+    implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+    {
+        if (!(node instanceof WindowNode)) {
+            return Optional.empty();
+        }
+
+        WindowNode parent = (WindowNode) node;
+
+        PlanNode source = lookup.resolve(parent.getSource());
+        if (!(source instanceof WindowNode)) {
+            return Optional.empty();
+        }
+
+        WindowNode child = (WindowNode) source;
+
+        if (!child.getSpecification().equals(parent.getSpecification()) && !dependsOn(parent, child)) {
+            return Optional.empty();
+        }
+
+        ImmutableMap.Builder<Symbol, WindowNode.Function> functionsBuilder = ImmutableMap.builder();
+        functionsBuilder.putAll(parent.getWindowFunctions());
+        functionsBuilder.putAll(child.getWindowFunctions());
+
+        return Optional.of(new WindowNode(
+                parent.getId(),
+                child.getSource(),
+                parent.getSpecification(),
+                functionsBuilder.build(),
+                parent.getHashSymbol(),
+                parent.getPrePartitionedInputs(),
+                parent.getPreSortedOrderPrefix()));
+    }
+
+    private static boolean dependsOn(WindowNode parent, WindowNode child)
```

Comment:
As I said https://github.com/Teradata/presto/pull/495#issuecomment-279704417, this copied method is a temporary change. (Also, this PR is against @maciejgrzybek's branch so that we can put together all code in one place).